esp-hal/src/clock/mod.rs

```diff
@@ -319,21 +319,34 @@ impl Clocks {
     /// [`crate::init()`].
     #[cfg(systimer)]
     pub(crate) fn xtal_freq() -> HertzU32 {
-        if let Some(clocks) = Self::try_get() {
-            clocks.xtal_clock
-        } else {
-            Self::measure_xtal_frequency().frequency()
+        if esp_config::esp_config_str!("ESP_HAL_CONFIG_XTAL_FREQUENCY") == "auto" {
+            if let Some(clocks) = Self::try_get() {
+                return clocks.xtal_clock;
+            }
         }
+
+        Self::measure_xtal_frequency().frequency()
     }
 }
 
 #[cfg(esp32)]
 impl Clocks {
     fn measure_xtal_frequency() -> XtalClock {
-        if RtcClock::estimate_xtal_frequency() > 33 {
-            XtalClock::_40M
+        if esp_config::esp_config_str!("ESP_HAL_CONFIG_XTAL_FREQUENCY") == "auto" {
+            if RtcClock::estimate_xtal_frequency() > 33 {
+                XtalClock::_40M
+            } else {
+                XtalClock::_26M
+            }
         } else {
-            XtalClock::_26M
+            const {
+                match esp_config::esp_config_str!("ESP_HAL_CONFIG_XTAL_FREQUENCY").as_bytes() {
+                    b"auto" => XtalClock::Other(0),
```

Comment:
shouldn't this be `unreachable!`?

Reply:
it can't be, const eval just ends up panicking if you configure anything else

Reply:
ah somehow missed `const`